File: tools/geofences-to-kml.js
import fs from "fs";
// example data geofences from https://your-traccar-domain/api/geofences or use fetch
const geofences = [
  {
    id: 546,
    name: "COP",
    description: null,
    area: "POLYGON ((-3.742736604580289 115.6425934049417, -3.742736555961199 115.64271185754588, -3.7429653968457814 115.64272080160285, -3.7431489320357323 115.64273111018622, -3.7433397536774486 115.6427447371434, -3.743498822259044 115.64274333982256, -3.743735606771379 115.64275138932655, -3.7439019269505334 115.64274322188884, -3.744028619378184 115.64273490408696, -3.744130866950302 115.6427328850462, -3.7441279198699284 115.64234113215413, -3.743228220340061 115.64235433098429, -3.7431187000700925 115.6423516891029, -3.7428161331079535 115.64236247030364, -3.7427377433026727 115.64236622574622, -3.742736604580289 115.6425934049417))",
    attributes: '{"StreetType":"COP","GroupName":"Line Hauling"}',
    calendarid: null,
    geotype: "Line Hauling",
    group_name: "KM 1",
  },
];

// Function to reverse the order of coordinates in a string
function reverseCoordinates(coordinatesString) {
  return coordinatesString
    .trim()
    .split(",")
    .map((coord) => coord.trim().split(" "))
    .map((pair) => pair.reverse().join(",")) // Reverse the order of coordinates
    .join(" "); // Combine the coordinates back into a string
}

// Helper function to parse the area into KML coordinates
function parseAreaToKML(area) {
  const coordinates = area
    .replace(/(LINESTRING|POLYGON|\(|\))/g, "") // Remove area type and parentheses
    .trim();

  return reverseCoordinates(coordinates); // Reverse the order of coordinates
}

// Function to convert geofences data to KML format
function convertToKML(geofences) {
  const kmlHeader = `<?xml version="1.0" encoding="UTF-8"?>\n<kml>\n  <Document>\n    <name>No Group</name>\n`;
  const kmlFooter = `  </Document>\n</kml>`;

  const placemarks = geofences
    .map((geofence) => {
      const coordinates = parseAreaToKML(geofence.area);
      return `
    <Placemark>
      <name>${geofence.name}</name>
      <description width="0.0" color="60DCDCDC" lineColor="FF027AFF"/>
      <Style>
        <PolyStyle>
          <color>60DCDCDC</color>
        </PolyStyle>
        <LineStyle>
          <color>FF027AFF</color>
        </LineStyle>
      </Style>
      <Polygon>
        <outerBoundaryIs>
          <LinearRing>
            <coordinates>${coordinates},0</coordinates>
          </LinearRing>
        </outerBoundaryIs>
      </Polygon>
    </Placemark>`;
    })
    .join("\n");

  return kmlHeader + placemarks + kmlFooter;
}

// Convert geofences data to KML format
const kmlData = convertToKML(geofences);

// Save the KML data to a file
fs.writeFileSync("geofences.kml", kmlData, "utf8"); // Write the KML data to a file
console.log("File KML has been created successfully.");

// how to run tools
// node /tools/geofences-to-kml.js

File: tools/geofences-to-kml.py
import os
import requests
import xml.etree.ElementTree as ET

BASE_URL = "http://localhost:3000/api"
KML_FILE_PATH = "geofences.kml"
ADMIN_EMAIL = "admin"
ADMIN_PASSWORD = "password"

session_cookie = ""

def login():
    """
    Login to the API and retrieve session cookie
    """
    global session_cookie
    login_data = {
        "email": ADMIN_EMAIL,
        "password": ADMIN_PASSWORD
    }
    headers = {"Content-Type": "application/x-www-form-urlencoded"}
    try:
        response = requests.post(f"{BASE_URL}/session", headers=headers, data=login_data)
        response.raise_for_status()
        session_cookie = response.headers.get("set-cookie").split(";")[0]
        print("Login successful, session cookie acquired.")
    except requests.RequestException as error:
        print("Error during login:", error)
        raise

def fetch_geofences():
    """
    Fetch geofences from the API
    """
    headers = {
        "Content-Type": "application/json",
        "Cookie": session_cookie
    }
    try:
        response = requests.get(f"{BASE_URL}/geofences", headers=headers)
        response.raise_for_status()
        print("Fetched geofences successfully.")
        return response.json()
    except requests.RequestException as error:
        print("Error fetching geofences:", error)
        return []

def reverse_coordinates(coordinates_string):
    """
    Reverse coordinates from "lat lon" to "lon,lat"
    """
    return " ".join(",".join(reversed(coord.strip().split(" "))) for coord in coordinates_string.strip().split(","))

def parse_area_to_kml(area):
    """
    Parse area string to KML-compatible coordinates
    """
    coordinates = area.replace("LINESTRING", "").replace("POLYGON", "").replace("(", "").replace(")", "").strip()
    return reverse_coordinates(coordinates)

def convert_to_kml(geofences):
    """
    Convert geofences data to KML format
    """
    kml = ET.Element("kml")
    document = ET.SubElement(kml, "Document")
    ET.SubElement(document, "name").text = "No Group"

    for geofence in geofences:
        placemark = ET.SubElement(document, "Placemark")
        ET.SubElement(placemark, "name").text = geofence["name"]

        style = ET.SubElement(placemark, "Style")
        poly_style = ET.SubElement(style, "PolyStyle")
        ET.SubElement(poly_style, "color").text = "60DCDCDC"
        line_style = ET.SubElement(style, "LineStyle")
        ET.SubElement(line_style, "color").text = "FF027AFF"

        polygon = ET.SubElement(placemark, "Polygon")
        outer_boundary = ET.SubElement(polygon, "outerBoundaryIs")
        linear_ring = ET.SubElement(outer_boundary, "LinearRing")

        coordinates = parse_area_to_kml(geofence["area"]) + ",0"
        ET.SubElement(linear_ring, "coordinates").text = coordinates

    return ET.tostring(kml, encoding="utf-8").decode("utf-8")

def write_kml_file(kml_data):
    """
    Write KML data to a file
    """
    try:
        with open(KML_FILE_PATH, "w", encoding="utf-8") as file:
            file.write(kml_data)
        print("KML file created successfully.")
    except IOError as error:
        print("Error writing KML file:", error)

def create_geofences_kml_file():
    """
    Main function to create the geofences KML file
    """
    try:
        login()
        geofences = fetch_geofences()
        kml_data = convert_to_kml(geofences)
        write_kml_file(kml_data)
    except Exception as error:
        print("Error in creating KML file:", error)

# Run the main function
if __name__ == "__main__":
    create_geofences_kml_file()
